contracts/Market.sol

```diff
@@ -0,0 +1,857 @@
+    address[] memory suppliers = new address[](numberOfActiveRemovalsInMarket);
+    uint256 numberOfRemovalsForOrder = 0;
+    // TODO (Gas Optimization): Declare variables outside of loop
+    for (
+      uint256 i = 0;
+      i < numberOfActiveRemovalsInMarket;
+      i++ // todo ++i consistency
+    ) {
+      uint256 removalId = _activeSupply[_currentSupplierAddress]
+        .getNextRemovalForSale();
+      uint256 removalAmount = _removal.balanceOf(address(this), removalId); // todo batch
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't increment currentSupplierAddress,
+         * don't check about removing active supplier.
+         */
+        ids[numberOfRemovalsForOrder] = removalId;
+        amounts[numberOfRemovalsForOrder] = remainingAmountToFill;
+        suppliers[numberOfRemovalsForOrder] = _currentSupplierAddress;
+        remainingAmountToFill = 0;
+      } else {
+        /**
+         * We will use up this removal while completing the order, move on to next one.
+         */
+        ids[numberOfRemovalsForOrder] = removalId;
+        amounts[numberOfRemovalsForOrder] = removalAmount; // this removal is getting used up
+        suppliers[numberOfRemovalsForOrder] = _currentSupplierAddress;
+        remainingAmountToFill -= removalAmount;
+        _activeSupply[_currentSupplierAddress].removeRemoval(removalId);
+        if (_activeSupply[_currentSupplierAddress].isRemovalQueueEmpty()) {
+          /**
+           * If the supplier is out of supply, remove them from the active suppliers.
+           */
+          _removeActiveSupplier(_currentSupplierAddress);
+        } else if (
+          /**
+           *  If the supplier is the only supplier remaining with supply, don't bother incrementing.
+           */
+          _suppliersInRoundRobinOrder[_currentSupplierAddress]
+            .nextSupplierAddress != _currentSupplierAddress
+        ) {
+          _incrementCurrentSupplierAddress();
+        }
+      }
+      numberOfRemovalsForOrder++;
+      if (remainingAmountToFill == 0) {
+        break;
+      }
+    }
+    // todo revert if total from suppliers != certificate amount
+    return (numberOfRemovalsForOrder, ids, amounts, suppliers);
+  }
+
+  /**
+   * @dev Determines the removal ids and amounts to fill the given purchase quantity, sourcing only
+   * from a single supplier.
+   *
+   * @param certificateAmount The number of carbon removals to purchase.
+   * @param supplier The supplier from which to purchase carbon removals.
+   * @return numberOfRemovalForOrder The number of distinct removal token ids used to fulfill this order.
+   * @return ids An array of the removal token ids being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   */
+  function _allocateSupplySingleSupplier(
```

Comment:
My suggestion above would remove most of the duplicated code here.

Reply:
tracked